test/lit/passes/minimize-rec-groups.wast

```diff
@@ -0,0 +1,486 @@
+  (global $g (ref null $t) (ref.null none))
+)
+
+;; Split a rec group containing independent types
+(module
+  (rec
+    ;; CHECK:      (type $a (struct (field i32)))
+    (type $a (struct (field i32)))
+    ;; CHECK:      (type $b (struct (field i64)))
+    (type $b (struct (field i64)))
+    ;; CHECK:      (type $c (struct (field f32)))
+    (type $c (struct (field f32)))
+  )
+
+  ;; CHECK:      (global $a (ref null $a) (ref.null none))
+  (global $a (ref null $a) (ref.null none))
+  ;; CHECK:      (global $b (ref null $b) (ref.null none))
+  (global $b (ref null $b) (ref.null none))
+  ;; CHECK:      (global $c (ref null $c) (ref.null none))
+  (global $c (ref null $c) (ref.null none))
+)
+
+;; Split a rec group containing types that depend on each other but belong to
+;; different SCCs.
+(module
+  (rec
+    ;; CHECK:      (type $a (struct))
+    (type $a (struct))
+    ;; CHECK:      (type $b (struct (field (ref $a))))
+    (type $b (struct (field (ref $a))))
+    ;; CHECK:      (type $c (struct (field (ref $b))))
+    (type $c (struct (field (ref $b))))
+  )
+
+  ;; CHECK:      (global $a (ref null $a) (ref.null none))
+  (global $a (ref null $a) (ref.null none))
+  ;; CHECK:      (global $b (ref null $b) (ref.null none))
+  (global $b (ref null $b) (ref.null none))
+  ;; CHECK:      (global $c (ref null $c) (ref.null none))
+  (global $c (ref null $c) (ref.null none))
+)
+
+;; Reverse the order of the previous case. The output should still be in a valid
+;; order.
+(module
+  (rec
+    ;; CHECK:      (type $a (struct))
+
+    ;; CHECK:      (type $b (struct (field (ref $a))))
+
+    ;; CHECK:      (type $c (struct (field (ref $b))))
+    (type $c (struct (field (ref $b))))
+    (type $b (struct (field (ref $a))))
+    (type $a (struct))
+  )
+
+  ;; CHECK:      (global $c (ref null $c) (ref.null none))
+  (global $c (ref null $c) (ref.null none))
+  ;; CHECK:      (global $b (ref null $b) (ref.null none))
+  (global $b (ref null $b) (ref.null none))
+  ;; CHECK:      (global $a (ref null $a) (ref.null none))
+  (global $a (ref null $a) (ref.null none))
+)
+
+;; Now all the types are in the same SCC.
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $c (struct (field (ref $a))))
+
+    ;; CHECK:       (type $b (struct (field (ref $c))))
+
+    ;; CHECK:       (type $a (struct (field (ref $b))))
+    (type $a (struct (field (ref $b))))
+    (type $b (struct (field (ref $c))))
+    (type $c (struct (field (ref $a))))
+  )
+
+  ;; CHECK:      (global $a (ref null $a) (ref.null none))
+  (global $a (ref null $a) (ref.null none))
+)
+
+;; Only two of the types are in the same SCC.
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $b (struct (field (ref $a))))
+
+    ;; CHECK:       (type $a (struct (field (ref $b))))
+    (type $a (struct (field (ref $b))))
+    (type $b (struct (field (ref $a))))
+    ;; CHECK:      (type $c (struct (field (ref $a))))
```

Comment:
More of a missing feature than a bug, I'd say, but yes. Getting the updater to print `(rec` at all requires some special casing since there is no `$name` there. I can take a look at how difficult it would be to get it to emit the closing paren, too.

Reply:
Ok I took a look at this but have given up because we are already several iterations into https://xkcd.com/1171/ with the auto update script for lit tests.